packages/babel-parser/src/util/scopeflags.js

```diff
@@ -41,12 +41,14 @@ export const BIND_NONE = 0, // Not a binding
   BIND_LEXICAL = 2, // Let- or const-style binding
   BIND_FUNCTION = 3, // Function declaration
   BIND_SIMPLE_CATCH = 4, // Simple (identifier pattern) catch binding
-  BIND_OUTSIDE = 5; // Special case for function names as bound inside the function
+  BIND_OUTSIDE = 5, // Special case for function names as bound inside the function
+  BIND_TS_ENUM = 6; // TypeScript enums (block-scoped, but can be re-declared using var/function)
```

Comment:
A complete fix for https://github.com/babel/babel/issues/9763 would also need to keep track of other typescript declarations.

Also, this is not quite how TS Enums work. `enum` can be declared multiple times, but only with `enum` or `namespace`. It might be useful to keep track of which types are in "type space" or "value space".

For example, `type A = {}; function A() {}` are completely different and unrelated -- they are two separate bindings (and both are hoisted!); they just share the same name.

Here's an attempt to explain how the declarations (are supposed to) work. The "type" part of declarations is always hoisted.

\ | value | type
--- | --- | ---
`namespace` | maybe | yes, can hold other declarations
`enum` | yes, `let`-like | yes, `namespace`-like
`class` | yes, `let`-like | yes (refers to the instance)
`interface` | no | yes
`type` | no | yes
`var` | yes | no
`let` | yes | no
`const` | yes | no
`function` | yes | no

Generic arguments aren't here but they are `type`-like.

`namespace` and `enum` (besides an ES module itself) are the only kinds of type that can hold other types or declarations inside them. `class` "looks like" it would behave like that but they are `interface`-like instead of `namespace`-like -- but `class` _can_ merge with a `namespace`, which affects the value portion `class` (i.e. they're part of the constructor, not the instance). `class` can also merge with `interface`, which affects the type portion of `class` (the instance type).

✅ = merges
🔴 = invalid
⚪️ = no merging _and_ no conflict; they are separate entities
⚠️ = namespaces are weird (_as long as_ they do not contain values they don't conflict; if they do they can't merge)
❗️ = no conflict if and only if the type annotations are identical

Merging | `namespace` | `enum` | `class` | `interface` | `type` | `var` | `let` | `const` | `function`
--- | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-: | :-:
`namespace` | ✅ | ✅ | ✅ | ✅ | 🔴 | ⚠ | ⚠ | ⚠️ | ✅  
`enum` | ✅ | ✅ | 🔴 | 🔴 | 🔴 | 🔴  | 🔴 | 🔴 | 🔴  
`class` | ✅ | 🔴 | 🔴 | ✅ | 🔴 | 🔴 | 🔴 | 🔴 | 🔴 | 🔴 
`interface` | ✅ | 🔴 | ✅ | ✅ | 🔴 | ⚪️ | ⚪️ | ⚪️ | ⚪️ 
`type` | 🔴 | 🔴 | 🔴 | 🔴 | 🔴 | ⚪️ | ⚪️ | ⚪️ | ⚪️ 
`var` | ⚠ | 🔴 | 🔴 | ⚪️ | ⚪️ | ❗️ | 🔴 | 🔴 | 🔴 | 🔴  
`let` | ⚠ | 🔴 | 🔴 | ⚪️ | ⚪️ | 🔴 | 🔴 | 🔴 | 🔴 | 🔴 
`const` | ⚠ | 🔴 | 🔴 | ⚪️ | ⚪️ | 🔴 | 🔴 | 🔴 | 🔴 | 🔴 
`function` | ✅ | 🔴 | 🔴 | ⚪️ | ⚪️ | 🔴 | 🔴 | 🔴 | `*`

`*`: multiple declarations of a function / method on the same scope are only valid if the declarations are overloads; only one of the declarations is allowed to have an implementation, and it must be the last one.

`type` can maybe be understood as the type-only `const` equivalent, as it is block-scoped and it can't merge with anything or be reassigned.

Reply:
Thanks you for the detailed table!
I think that we for the (`var`...`functon`)x(`var`...`functon`) cells, we should follow JavaScript's behavior, since we can't rely on type info in Babel.

As for :warning:, I think that I won't implement redeclaration checks for them in this PR since they are only needed when transpiling namespaces to IIFEs, but Babel doesn't support doing so. It could be added later but it will be more difficult.

Reply:
> The "type" part of declarations is always hoisted.

Always to the block, not to the function, right?